src/fido2_child/fido2_child_common.c

```diff
@@ -0,0 +1,302 @@
+        ret = ENOMEM;
+        goto done;
+    }
+
+    if (debug_fd != -1) {
+        opt_logger = sss_logger_str[FILES_LOGGER];
+        ret = set_debug_file_from_fd(debug_fd);
+        if (ret != EOK) {
+            opt_logger = sss_logger_str[STDERR_LOGGER];
+            ERROR("set_debug_file_from_fd failed.\n");
+        }
+    }
+
+    DEBUG_INIT(debug_level, opt_logger);
+
+    ret = EOK;
+
+done:
+    return ret;
+}
+
+errno_t
+check_arguments(const struct fido2_data *data)
+{
+    errno_t ret = EOK;
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Argument values after parsing\n");
+    DEBUG(SSSDBG_TRACE_FUNC, "action: %d\n", data->action);
+    DEBUG(SSSDBG_TRACE_FUNC, "shortname: %s, domain: %s\n",
+          data->shortname, data->domain);
+    DEBUG(SSSDBG_TRACE_FUNC, "public_key: %s, key_handle: %s\n",
+          data->public_key, data->key_handle);
+    DEBUG(SSSDBG_TRACE_FUNC, "type: %d\n", data->type);
+    DEBUG(SSSDBG_TRACE_FUNC, "user_verification: %d\n",
+          data->user_verification);
+    DEBUG(SSSDBG_TRACE_FUNC, "debug_libfido2: %d\n", data->debug_libfido2);
+
+    if (data->action == ACTION_NONE) {
+        DEBUG(SSSDBG_OP_FAILURE, "No action set.\n");
+        ret = ERR_INPUT_PARSE;
+        goto done;
+    }
+
+    if (data->action == ACTION_REGISTER
+        && (data->shortname == NULL || data->domain == NULL)) {
+        DEBUG(SSSDBG_OP_FAILURE, "Too few arguments for register action.\n");
+        ret = ERR_INPUT_PARSE;
+        goto done;
+    }
+
+    if (data->action == ACTION_AUTHENTICATE
+        && (data->shortname == NULL || data->domain == NULL
+        || data->public_key == NULL || data->key_handle == NULL)) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Too few arguments for authenticate action.\n");
+        ret = ERR_INPUT_PARSE;
+        goto done;
+    }
+
+done:
+    return ret;
+}
+
+errno_t
+register_key(struct fido2_data *data)
+{
+    fido_cred_t *cred = NULL;
+    fido_dev_t *dev = NULL;
+    fido_dev_info_t *dev_list = NULL;
+    size_t dev_number = 0;
+    errno_t ret;
+
+    cred = fido_cred_new();
+    if (cred == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_cred_new failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = prepare_credentials(data, cred);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    dev_list = fido_dev_info_new(DEVLIST_SIZE);
+    if (dev_list == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_dev_info_new failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = list_devices(dev_list, &dev_number);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    if (dev_number == 0) {
+        DEBUG(SSSDBG_OP_FAILURE, "No device found. Aborting.\n");
+        ret = ENOENT;
+        goto done;
+    }
+
+    dev = fido_dev_new();
+    if (dev == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_dev_new failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = select_device(dev_list, 0, dev);
```

Comment:
> This will always select the first device that happens to be enumerated by `fido_dev_info_manifest()`.
> 
> As a future improvement, consider letting a user with multiple authenticators plugged in select which one to use without having to unplug all others. A ~two~ way to do this is to:

I had already noted this and my idea is to document the limitation.

>     * Leverage `fido_dev_get_touch_{begin,status}()` to allow the user to select which authenticator to create the credential on via a touch (see `examples/select.c` in libfido2). This will require the user to touch their selected authenticator twice.

From a user perspective what would happen? Will they see all the keys blinking at the same time? Or do they blink one by one until a key is touched?

Reply:
> I had already noted this and my idea is to document the limitation.

Sounds good :+1:
 
> From a user perspective what would happen? Will they see all the keys blinking at the same time? Or do they blink one by one until a key is touched?

Depends on how the API is used. In the libfido2 example, all keys start to blink at the same time until one of them is touched.

Reply:
Hi,

what are your plan with `fido_dev_get_touch_{begin,status}()` if there is only one device connected? I guess in this case it would make sense to skip it and not ask first to select one by touching on it and then ask to touch it a second time. If you agree I think it would make sense to error out here if more than one device is found with a message that currently only a single device is supported. Then you can replace the error message with the selection code in some later patch.

bye,
Sumit

Reply:
>  I guess in this case it would make sense to skip it and not ask first to select one by touching on it and then ask to touch it a second time.

Agreed. FWIW, this is how most other clients (ssh, web browsers, ...) behave (including the aforementioned libfido2 example).

Reply:
I added the check to abort if more than one key is detected.